docs/proposals/201909_bucketfile.md

```diff
@@ -0,0 +1,287 @@
+
+### Combine queries
+
+```
+type BucketFile struct {
+    // ...
+    pendingDuration time.Duration   // duration to fetch pending pages
+    pendingPages    *roaring.Bitmap // pending pages to be fetched
+    pendingReaders  int             // record pending callers
+    pendingChan     chan error      // chan to notify callers the result of prefetch	
+}
+
+func (f *BucketFile) Prefetch(ranges []*Range) (chan error, error) {...}
+```
+
+BucketFile appends new pages to `pendingPages`, waits for next time to fetch. So that we can combine multiple queries to ones.
+
+If the combined prefetch fails, all sub prefetches will fail. There must be error handling for failed prefetches.
+
+Each time to fetch pending pages, BucketFile will reset `pendingPages`, `pendingReaders` and `pendingChan`.
+
+### In-memory objects
+
+Each IndexReader decodes [postings offset block](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#postings-offset-table) and [label offset block](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#label-offset-table) into in-memory object for sequential and random accesses.
+
+Each BucketFile has a `pages` bitmap, but the bitmap uses small memory. 
+
+For inactive IndexReaders and BucketFiles, if the inactive duration (`now.Sub(lastQueryTime)`) is greater than a configured duration (e.g. 15 mins), then IndexReaders will remove in-memory objects, and BucketFiles will remove in-memory bitmaps, close file descriptors and unmmap local files. This will reduce memory usage.
+
+### Preload blocks on startup
+
+The latest blocks will be queried the most. 
+
+Preload `the last N hours` (configurable) blocks on startup.
+This will fetch index files from object storage and decode bytes into in-memory objects.
+
+### Local files
+
+Because the max size of a index file is 64 GiB, and the default size of a chunks file is 512 MiB.
+
+So we can set the size of a local index file to 64 GiB, set the size of a local chunks file to 768 (512*1.5) MiB.
+
+The size is used to create a sparse file and mmap a file.
+
+### Object storage byte ranges
+
+Currently, different provider clients implement byte ranges differently. We need to wrap their implementations and export the same api.
+
+ I think byte ranges in [minio-go](https://github.com/minio/minio-go/blob/master/api-get-options.go#L99-L128) is a good example:
+
+1. Fetch last N bytes:     
+     `Range{start:0, end: -N}` ->  Request header`Range: bytes=-N`     
+     **NOTE**: will not write to local file because we do not know the offset.
+2. Fetch everything starting from `start`:    
+    `Range{start:N, end: 0}	`  ->  Request heade `Range: bytes=N-` 
+
+3. Fetch everything starting at `start` till  `end`:    
+   `Range{start:N, end: M}` -> Request header `Range: bytes=M-N` 
+
+Now all providers (Azure, COS, GCS, S3, Swift) support `Fetch last N bytes`, `Fetch everything starting from offset` and `Fetch everything starting at start till end`.
```

Comment:
Thanks for research, let's see first if we really need to have this working, maybe we can do it without those 3 special features of fetching ranges? 
Can you provide motivation here for them?

Reply:
Essentially, index TOC knows exactly where things are in the first place, so we need just that :thinking:

Reply:
we do not know the size of the remote index file, so we need the Feature-1 `Fetch last N bytes` to fetch TOC.

we know the start of `postings offset table` block through [TOC](https://github.com/prometheus/tsdb/blob/master/docs/format/index.md#toc), but we do not know the end of this block, so we need the Feature-2 `Fetch everything starting from start` to fetch TOC.

Reply:
TODO: check overfetch.

Reply:
> we know the start of postings offset table block through TOC, but we do not know the end of this block, so we need the Feature-2 Fetch everything starting from start to fetch TOC.

yes, you are right here.